renku/core/migration/m_0010__metadata_fixes.py

```diff
@@ -255,24 +251,38 @@ def fix_plan_times(activity_gateway: IActivityGateway, plan_gateway: IPlanGatewa
             plan.date_removed = None
 
         if plan.date_removed is not None:
-            if plan.date_removed < plan.date_created:
-                # NOTE: Fix invalidation times set before creation date on plans
-                plan.date_removed = plan.date_created
             if plan.date_removed.tzinfo is None:
                 # NOTE: There was a bug that caused date_removed to be set without timezone (as UTC time)
                 # so we patch in the timezone here
                 plan.date_removed = plan.date_removed.replace(microsecond=0).astimezone(timezone.utc)
-        if plan.id in activity_map and plan.date_created > activity_map[plan.id].started_at_time:
-            plan.date_created = activity_map[plan.id].started_at_time
+            if plan.date_removed < plan.date_created:
+                # NOTE: Fix invalidation times set before creation date on plans
+                plan.date_removed = plan.date_created
         plan.freeze()
 
     fix_plan_dates(plans=plans, plan_gateway=plan_gateway)
-    database.commit()
+    project_context.database.commit()
+
+
+@inject.autoparams("activity_gateway")
+def fix_activity_times(activity_gateway: IActivityGateway):
+    """Make sure activity have valid start/end/delete dates."""
+    fix_activity_dates(activities=activity_gateway.get_all_activities(include_deleted=True))
+    project_context.database.commit()
 
 
 @inject.autoparams("dataset_gateway")
 def fix_dataset_date_modified(dataset_gateway: IDatasetGateway):
     """Change date_created and date_modified to have correct semantics."""
+
+    def fix_creation_date(dataset):
+        """Check creation date to make sure that it's after project's creation date."""
+        if dataset.date_created and dataset.date_created < project_context.project.date_created:
+            try:
+                dataset.date_created = min([f.date_added for f in dataset.files])
```

Comment:
Should we check that dataset.date_created is >= project.date_created after fixing it here, in case a dataset file also had a wrong date?